src/parse.rs

```diff
@@ -0,0 +1,320 @@
+                let token_id: usize = lexeme.tok_id().try_into().ok().unwrap();
+                grm.term_name(TIdx::from(token_id)).unwrap().to_string()
+            }
+        }
+    };
+
+    fn gen_class_body(node: &Node<u16>, grm: &YaccGrammar, input: &str, class: &mut Class) {
+        if let &Node::Nonterm { nonterm_idx, ref nodes } = node {
+            for child in nodes.iter() {
+                if let &Node::Nonterm { nonterm_idx, ref nodes} = child {
+                    let name = term_name(child, grm, input);
+                    match name.as_ref() {
+                        "method_def" => gen_method(child, grm, input, class),
+                        "class_body" => gen_class_body(child, grm, input, class),
+                        _ => (),
+                    }
+                }
+            }
+        }
+    }
+
+    // method_def : "DEF" "IDENTIFIER" "LPAREN" parameter_list_opt "RPAREN" block ;
+    fn gen_method(node: &Node<u16>, grm: &YaccGrammar, input: &str, class: &mut Class) {
+        if let &Node::Nonterm{ nonterm_idx, ref nodes } = node {
+            let mut method = Method::new();
+            let meth_name = id_value(&nodes[1], grm, input);
+            gen_params(&nodes[3], grm, input, &mut method);
+            gen_block(&nodes[5], grm, input, &mut method);
+            class.add_method(meth_name, method)
+        }
+    }
+
+    fn gen_params(node: &Node<u16>, grm: &YaccGrammar, input: &str, method: &mut Method) {
+        match *node {
+            Node::Nonterm { nonterm_idx, ref nodes } => {
+                for child in nodes.iter() {
+                    gen_params(child, grm, input, method)
+                }
+            }
+            Node::Term{ lexeme } => {
+                if term_name(node, grm, input) == "IDENTIFIER" {
+                    let var_name = id_value(node, grm, input);
+                    method.push_var(var_name);
+                }
+            }
+        }
+    }
+
+    fn gen_block(node: &Node<u16>, grm: &YaccGrammar, input: &str, method: &mut Method) {
+        if let &Node::Nonterm{ nonterm_idx, ref nodes } = node {
+            for child in nodes {
+                let name = term_name(child, grm, input);
+                match name.as_ref() {
+                    "statement" => gen_stmt(child, grm, input, method),
+                    _ => gen_block(child, grm, input, method)
+                }
+            }
+        }
+    }
+
+    // statement : expression
+    //           | if_statement
+    //           | let_statement
+    //           | for_statement
+    //           ;
+    fn gen_stmt(node: &Node<u16>, grm: &YaccGrammar, input: &str, method: &mut Method) {
+        if let &Node::Nonterm{ nonterm_idx, ref nodes } = node {
+            let name = term_name(&nodes[0], grm, input);
+            println!("{}", name);
+            match name.as_ref() {
+                "expression" => gen_exp(&nodes[0], grm, input, method),
+                "if_statement" => (),
```

Comment:
I assume this means "I haven't got around to `if_statement` yet"? If so, a `panic` would be much safer (at least you'll know, if you use an `if`, that something's missing, rather than it seeming to work but actually doing nothing).

Reply:
Fixed - ish (I do have if and for working now but I think it's best to push them as part of the PR to stop this one becoming too bloated)

Reply:
I'd still like this to be a `panic` in this PR. You can then change the `panic` into something nicer in a future PR.